### PR TITLE
support data parallel + pipeline parallel

### PR DESCRIPTION
**Pull request by xzhu1900**
**Description**: Describe your changes.
This change adds support to run pipeline parallel together with data parallel. Tested with Bert-L and the corresponding unit tests are added. 

**Motivation and Context**
- Why is this change required? What problem does it solve?
Previously pipeline parallel and data parallel are incompatible. Ort can only run one or the other. Now with a node of m x n GPUs, it can run m data parallel and n pipeline parallel. 

The command to run Bert-Large is as below:

/bert_ort/openmpi/bin/mpirun --report-pid - -n 4 ./onnxruntime_training_bert --model_name /bert_ort/bert_models/nv/bert-large/bert-large-uncased_L_24_H_1024_A_16_V_30528_S_512_Dp_0.1_optimized_layer_norm_opset12 --train_data_dir /bert_data/512/books_wiki_en_corpus/train --test_data_dir /bert_data/512/books_wiki_en_corpus/test --train_batch_size 1 --mode train --display_loss_steps 1 --optimizer lamb --learning_rate 0.006 --gradient_accumulation_steps 5 --num_train_steps 10 --warmup_ratio 0 --warmup_mode Linear --use_nccl --pipeline_parallel_size 2 --cut_group_info 1881:407-1951/2073/2195/2317/2439/2561/2683/2805/2927/3049/3171/3293 --data_parallel_size 2

- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: orttraining/orttraining/core/framework/distributed_run_context.cc [hunk 1]
```diff
@@ -47,14 +45,46 @@ DistributedRunContext::DistributedRunContext(int32_t world_rank,
   params_.data_parallel_size = data_parallel_size;
   params_.horizontal_parallel_size = horizontal_parallel_size;
   params_.pipeline_stage_size = pipeline_stage_size;
-  groups_.resize(2);
+  groups_.resize(static_cast<size_t>(WorkerGroupType::WorkerGroupTypeCount));
 
+  // Consider distributed training forms three axis, data parallel, horizontal parallel and pipeline parallel.
+  // The three axis are pependicular to each other and like x, y, z axis in 3D space (but only the positive directions).
```
Comment (wschin) [inline orttraining/orttraining/core/framework/distributed_run_context.cc:51]: >axis [](start = 15, length = 4)

axis->axes (same below)

File: orttraining/orttraining/core/framework/distributed_run_context.cc [hunk 2]
```diff
@@ -47,14 +45,46 @@ DistributedRunContext::DistributedRunContext(int32_t world_rank,
...
+  //            |__________________> data
+  //             \ 0, 4, 8,
+  //              \ 1, 5, 9,
+  //               \ 2, 6, 10,
+  //                \ 3, 7, 11,
+  //                 v
+  //                horizontal
+  //
+  // For a given world rank, say 11, it will be in the 3th data parallel group (3, 7, 11), with in-group index 2,
+  // and be in the 2th horizontal parallel group (8, 9, 10, 11) with in-group index 3; and be in the 11th pipeline
```
Comment (wschin) [inline orttraining/orttraining/core/framework/distributed_run_context.cc:74]: >2th [](start = 19, length = 3)

Isn't it the 4th horizontal parallel graph containing (3, 7, 11)?

File: orttraining/orttraining/core/framework/distributed_run_context.cc [hunk 3]
```diff
@@ -47,14 +45,46 @@ DistributedRunContext::DistributedRunContext(int32_t world_rank,
...
+  //            |    15, 19, 23,
+  //            |__________________> data
+  //             \ 0, 4, 8,
+  //              \ 1, 5, 9,
+  //               \ 2, 6, 10,
+  //                \ 3, 7, 11,
+  //                 v
+  //                horizontal
+  //
+  // For a given world rank, say 11, it will be in the 3th data parallel group (3, 7, 11), with in-group index 2,
```
Comment (wschin) [inline orttraining/orttraining/core/framework/distributed_run_context.cc:73]: >(3, 7, 11) [](start = 79, length = 10)

Not sure if it's (8, 9, 10, 11)?

File: orttraining/orttraining/core/framework/distributed_run_context.cc [hunk 4]
```diff
@@ -47,14 +45,46 @@ DistributedRunContext::DistributedRunContext(int32_t world_rank,
...
+  // parallel group (11, 23), with in-group index 0. Both the group id and in-group index are 0-based indexing.
+  //
+  // The calculation below is for a given world_rank, calculating its group index, in-group index and all ranks in this
+  // particular group.
+  //
   // Initialize Data Parallel Group
-  const int32_t data_group_id = world_rank % horizontal_parallel_size;
-  const int32_t rank_in_owning_data_group = world_rank / horizontal_parallel_size;
+  const int32_t data_group_id = world_rank / horizontal_parallel_size / data_parallel_size * horizontal_parallel_size + world_rank % horizontal_parallel_size;
+  const int32_t a = world_rank / horizontal_parallel_size / data_parallel_size * horizontal_parallel_size * data_parallel_size + world_rank % horizontal_parallel_size;
```
Comment (wschin) [inline orttraining/orttraining/core/framework/distributed_run_context.cc:82]: >horizontal_parallel_size / data_parallel_size * horizontal_parallel_size [](start = 33, length = 72)

divided by horizontal_parallel_size and multiplied by horizontal_parallel_size?

File: orttraining/orttraining/core/framework/distributed_run_context.cc [hunk 5]
```diff
@@ -47,14 +45,46 @@ DistributedRunContext::DistributedRunContext(int32_t world_rank,
...
+  // and be in the 2th horizontal parallel group (8, 9, 10, 11) with in-group index 3; and be in the 11th pipeline
+  // parallel group (11, 23), with in-group index 0. Both the group id and in-group index are 0-based indexing.
+  //
+  // The calculation below is for a given world_rank, calculating its group index, in-group index and all ranks in this
+  // particular group.
+  //
   // Initialize Data Parallel Group
-  const int32_t data_group_id = world_rank % horizontal_parallel_size;
-  const int32_t rank_in_owning_data_group = world_rank / horizontal_parallel_size;
+  const int32_t data_group_id = world_rank / horizontal_parallel_size / data_parallel_size * horizontal_parallel_size + world_rank % horizontal_parallel_size;
```
Comment (wschin) [inline orttraining/orttraining/core/framework/distributed_run_context.cc:81]: >horizontal_parallel_size [](start = 93, length = 24)

Removed?

File: orttraining/orttraining/core/framework/distributed_run_context.cc (hunk 3 above)
Comment (xzhu1900) [reply orttraining/orttraining/core/framework/distributed_run_context.cc:73]: numbers aligned with horizontal axis is the horizontal group. So (8, 9, 10, 11) form a horizontal group. (It matches the horizontal parallel size 4 too.)

File: orttraining/orttraining/core/framework/distributed_run_context.cc (hunk 2 above)
Comment (xzhu1900) [reply orttraining/orttraining/core/framework/distributed_run_context.cc:74]: (3, 7, 11) is the 3th data parallel group (counting start from 0). Ranks in a horizontal parallel group split the weights of certain tensors in GPU.

File: orttraining/orttraining/core/framework/distributed_run_context.cc (hunk 3 above)
Comment (wschin) [reply orttraining/orttraining/core/framework/distributed_run_context.cc:73]: Are GPUs with different horizontal coordinate in the same horizontal group? In this coordinate system, (3, 7, 11) have the same horizontal coordinate but different data coordinates, so I feel (3, 7, 11) is horizontal group.

---
In reply to: [462639398](https://github.com/microsoft/onnxruntime/pull/4648#discussion_r462639398) [](ancestors = 462639398)

Comment (xzhu1900) [reply orttraining/orttraining/core/framework/distributed_run_context.cc:73]: Nah, as mentioned in the comment, ranks that are aligned with a certain axis forms a group in that particular parallelism. It is not related to coordinates. (Also, coordinate is not a good way to describe groups in this case, as (15, 19, 23) also has the same horizontal coordinate as (3, 7, 11) in this example.) So the best way to depict groups is viewing them as axis-aligned groups.

File: orttraining/orttraining/core/framework/distributed_run_context.cc (hunk 5 above)
Comment (xzhu1900) [reply orttraining/orttraining/core/framework/distributed_run_context.cc:81]: yeah it also took me a while to figure out the existing system. One way to verify it is correct is through tests. DistributedRunContextTest is quite extensive. Apart from the newly added tests, the existing ones for D and H are none-modified and are passing, which gives me some confidence that this equation is right. 

Let me know what part that is still puzzling, or we can sync offline on this too.

File: orttraining/orttraining/core/framework/distributed_run_context.cc [hunk 6]
```diff
@@ -47,14 +45,46 @@ DistributedRunContext::DistributedRunContext(int32_t world_rank,
...
+  // The three axis are pependicular to each other and like x, y, z axis in 3D space (but only the positive directions).
+  // Now the world ranks are numbers filling in this 3D space alone the three axis. It will try fill alone the
+  // horizontal axis first, then data parallel axis, and last pipeline parallel axis.
+  //
+  // Ranks that are aligned with a specific axis forms a group. For example, for a 3D cube with size
+  // horizontal_parallel x data_parallel x pipeline_parallel equal to 4x3x2, there are 24 rank numbers fill into the
+  // cubic. It will have 6 horizontal groups, 8 data parallel groups and 12 pipeline groups, as shown below.
+  //
+  //          pipeline
+  //            ^
```
Comment (wschin) [inline orttraining/orttraining/core/framework/distributed_run_context.cc:60]: For machine ID c, are the following equations correct?

x (in_group_hori_group_id) = c % 4
y (in_group_data_group_id) = (c / 4) % 3
z (in_group_pipe_group_id) = (c / 12) % 2

To recover linear index x, we can use x + y * 4 + z * 4 * 3.
To see which hori group x belongs to, we can reuse the linear-index equation but assume there is no hori axis by
 1. set dimension of hori group to 1 (instead of 4)
 2. set  in_group_hori_group_id to 0.
hori_group_id = 0 +y * 1 +z * 1 * 3
Similarly, we have
data_group_id = x + 0 * 3 + z * 4 * 1
pipe_group_id = x + y * 4 + 0 * 4 * 3

A more systematic understanding is to think this as calculating linear index in the projected space.

Comment (xzhu1900) [reply orttraining/orttraining/core/framework/distributed_run_context.cc:60]: This is neat. Thanks. Update. 

---
In reply to: [478766878](https://github.com/microsoft/onnxruntime/pull/4648#discussion_r478766878) [](ancestors = 478766878)

File: orttraining/orttraining/core/framework/distributed_run_context.cc (hunk 1 above)
Comment (xzhu1900) [reply orttraining/orttraining/core/framework/distributed_run_context.cc:51]: Fixed.

---
In reply to: [462631388](https://github.com/microsoft/onnxruntime/pull/4648#discussion_r462631388) [](ancestors = 462631388)

File: orttraining/orttraining/core/framework/distributed_run_context.h [hunk 7]
```diff
@@ -66,6 +78,14 @@ class DistributedRunContext {
     return DistributedRunContext::GetInstance().GetWorkerGroup(group_type).rank_in_group;
   }
 
+  static int32_t GroupId(WorkerGroupType group_type) {
+    return DistributedRunContext::GetInstance().GetWorkerGroup(group_type).group_id;
+  }
+
+  static std::vector<int32_t> GetPipelineRanks(){
```
Comment (wschin) [inline orttraining/orttraining/core/framework/distributed_run_context.h:85]: >GetPipelineRanks [](start = 30, length = 16)

Consider
```
static std::vector<int32_t> GetRanks(WorkerGroupType group_type)
```
following the style of the above two functions.

File: orttraining/orttraining/core/session/training_session.cc [hunk 8]
```diff
@@ -152,13 +152,28 @@ Status TrainingSession::ConfigureForTraining(
                                          config.distributed_config.horizontal_parallel_size,
                                          config.distributed_config.pipeline_parallel_size});
 
+  int32_t pipeline_stage_id = config.pipeline_config.has_value() ?
```
Comment (wschin) [inline orttraining/orttraining/core/session/training_session.cc:155]: const int32_t?

File: orttraining/orttraining/core/framework/distributed_run_context.h [hunk 9]
```diff
@@ -34,6 +39,13 @@ struct DistributedRunConfig {
   int32_t pipeline_stage_size{1};
 };
```
Comment (wschin) [inline orttraining/orttraining/core/framework/distributed_run_context.h:40]: What are those fields valid ranges? Are their NULL values -1?

File: orttraining/orttraining/core/framework/distributed_run_context.cc [hunk 10]
```diff
@@ -47,27 +45,87 @@ DistributedRunContext::DistributedRunContext(int32_t world_rank,
...
+                                   const int32_t data_parallel_size,
+                                   const int32_t xx,
+                                   const int32_t yy,
+                                   const int32_t zz) {
+    return xx + yy * hori_parallel_size + zz * hori_parallel_size * data_parallel_size;
+  };
+
+  // Calculate the id of a group the current rank belongs to.
+  const int32_t hori_group_id = calculate_linear_index(1, data_parallel_size, 0, y, z);
+  const int32_t data_group_id = calculate_linear_index(horizontal_parallel_size, 1, x, 0, z);
```
Comment (wschin) [inline orttraining/orttraining/core/framework/distributed_run_context.cc:96]: Really beautiful.

File: orttraining/orttraining/models/runner/training_runner.cc [hunk 11]
```diff
@@ -21,6 +21,7 @@
 #include "orttraining/models/runner/training_util.h"
 #include "single_include/nlohmann/json.hpp"
 #include "test/perftest/utils.h"
+#include "orttraining/core/framework/distributed_run_context.h"
```
Comment (wschin) [inline orttraining/orttraining/models/runner/training_runner.cc:25]: move this up a little bit.

Comment (xzhu1900) [reply orttraining/orttraining/models/runner/training_runner.cc:25]: updated.

---
In reply to: [480302835](https://github.com/microsoft/onnxruntime/pull/4648#discussion_r480302835) [](ancestors = 480302835)

File: orttraining/orttraining/core/framework/distributed_run_context.h (hunk 9 above)
Comment (xzhu1900) [reply orttraining/orttraining/core/framework/distributed_run_context.h:40]: It ranges from 1 to world_size. And the values are validated in DistributedRunContext::DistributedRunContext()

Not sure what you mean by NULL values.. The default value is listed here and -1 should not be allowed (and will error out in the validation)

---
In reply to: [480287661](https://github.com/microsoft/onnxruntime/pull/4648#discussion_r480287661) [](ancestors = 480287661)

File: orttraining/orttraining/core/session/training_session.cc (hunk 8 above)
Comment (xzhu1900) [reply orttraining/orttraining/core/session/training_session.cc:155]: Thanks. Updated. 

---
In reply to: [479841534](https://github.com/microsoft/onnxruntime/pull/4648#discussion_r479841534) [](ancestors = 479841534)

File: orttraining/orttraining/core/framework/distributed_run_context.h (hunk 9 above)
Comment (wschin) [reply orttraining/orttraining/core/framework/distributed_run_context.h:40]: Sorry. I mean ID value such as
```
  const int32_t pipeline_stage_id = config.pipeline_config.has_value() ?
                              DistributedRunContext::RankInGroup(WorkerGroupType::ModelParallel) :
                              -1;
```
but looks like they are not IDs.

---
In reply to: [480409920](https://github.com/microsoft/onnxruntime/pull/4648#discussion_r480409920) [](ancestors = 480409920,480287661)
